hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterClientProtocol.java

```diff
@@ -667,39 +667,28 @@ public void rename2(final String src, final String dst,
   public void concat(String trg, String[] src) throws IOException {
     rpcServer.checkOperation(NameNode.OperationCategory.WRITE);
 
-    // See if the src and target files are all in the same namespace
-    LocatedBlocks targetBlocks = getBlockLocations(trg, 0, 1);
-    if (targetBlocks == null) {
-      throw new IOException("Cannot locate blocks for target file - " + trg);
-    }
-    LocatedBlock lastLocatedBlock = targetBlocks.getLastLocatedBlock();
-    String targetBlockPoolId = lastLocatedBlock.getBlock().getBlockPoolId();
-    for (String source : src) {
-      LocatedBlocks sourceBlocks = getBlockLocations(source, 0, 1);
-      if (sourceBlocks == null) {
-        throw new IOException(
-            "Cannot located blocks for source file " + source);
-      }
-      String sourceBlockPoolId =
-          sourceBlocks.getLastLocatedBlock().getBlock().getBlockPoolId();
-      if (!sourceBlockPoolId.equals(targetBlockPoolId)) {
-        throw new IOException("Cannot concatenate source file " + source
-            + " because it is located in a different namespace"
-            + " with block pool id " + sourceBlockPoolId
-            + " from the target file with block pool id "
-            + targetBlockPoolId);
-      }
-    }
+    // Concat only effects when all files in same namespace.
+    // And in router view, a file only exists in one RemoteLocation.
```

Comment:
Normally a file only exits in one Namespace. But maybe multiple namespaces contain this file and RBF returns the file in the first namespace to the client.

Reply:
For the empty file, maybe we can use `getFileInfo` to get the namespace that this trg belongs to.

Reply:
Thanks for review.

> you can refer to this case:
> 
> * /path mounts to NS1, NS2 and NS3
> * NS2 and NS3 contains /path
> * `OrderedResolver` returns NS1, NS2 and NS3
> 
> for this case, we should proxy this `concat` to NS2 instead of NS1, right?

This only happens when /path is a directory. And for a file ,it must only be from one exactly nameservice via dfsrouter.

Reply:
![image](https://github.com/apache/hadoop/assets/6347715/f787e314-6f5e-41b2-9361-df5564e9fe52)
DFSRouter with different order(HASH,LOCAL, RANDOM, HASH_ALL, SPACE) , a file cannot be written into two or more nameservices.

Reply:
Emmm, there is a scene .


1) If a file is already exist in two nameservices. And the add router mount.
NS1     /user/test/file
NS2    /user/test/file

2)  Add router mount.
hdfs dfsrouteradmin -add /user/test NS1,NS2 /user/test  -order RANDOM

3)   getDestination
hdfs dfsrouteradmin -getDestination /user/test/file 
Will return  NS1,NS2

For a file in Router view with more than one nameservices, I think should thrown Exception for concat method.  
 Look forward to your guidance , thanks ! @ZanderXu

Reply:
> For a file in Router view with more than one nameservices, I think should thrown Exception for concat method.

For multiple namespaces contain one same file case, RBF just return the file in the first namespace currently, such as: getBlockLocation, getFileInfo, etc. 

So If you want to thrown Exception for concat, maybe you need to modify all RPCs to throw Exception for this case.

May we can use getFileInfo instead of getBlockLocation to fix this bug. BTW, getFileInfo is no needed if this path only mounts to one namespace.

Reply:
>May we can use getFileInfo instead of getBlockLocation to fix this bug. BTW, getFileInfo is no needed if this path only mounts to one namespace.

At the beginning , I consider use getFileInfo .  But the problem is that HdfsFileStatus does not have any information about nameservices or blockpoolid, only LocatedBlock has.

Reply:
we don't need to get NSId or BPId from the result of getFileInfo, you can refer to `invokeSequential` to loop all namespaces one by one.

1. Get all namespaces of the input path.
2. Send getFileInfo to each namespace one by one 
3. The first namespace that the result of getFileInfo is not null is the one we need